perf(ivy): ngcc - exit early if the targeted package has been compiled

petebacondarwin opened the pull request:
**This is a quick fix for webpack but there are more optimisations to be achieved on the
first time run by replacing the TS module resolver for the initial walk of the tree.**

Previously we always walked the whole folder tree looking for
entry-points before we tested whether a target package had been
processed already. This could take >10secs!

This commit does a quick check of the target package before doing
the full walk which brings down the execution time for ngcc in this
case dramatically.

```
$ time ./node_modules/.bin/ivy-ngcc -t @angular/common/http/testing
Compiling @angular/core : fesm2015 as esm2015
Compiling @angular/core : fesm5 as esm5
Compiling @angular/core : esm2015 as esm2015
Compiling @angular/core : esm5 as esm5
Compiling @angular/common/http : fesm2015 as esm2015
Compiling @angular/common/http : fesm5 as esm5
Compiling @angular/common/http : esm2015 as esm2015
Compiling @angular/common/http : esm5 as esm5
Compiling @angular/common/http/testing : fesm2015 as esm2015
Compiling @angular/common/http/testing : fesm5 as esm5
Compiling @angular/common/http/testing : esm2015 as esm2015
Compiling @angular/common/http/testing : esm5 as esm5

real	0m19.766s
user	0m28.533s
sys	0m2.262s
```

```
$ time ./node_modules/.bin/ivy-ngcc -t @angular/common/http/testing
The target entry-point has already been processed

real	0m0.666s
user	0m0.605s
sys	0m0.113s
```

alan-agius4 commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -179,4 +185,11 @@ function markNonAngularPackageAsProcessed(path: AbsoluteFsPath, propertiesToCons
+function hasProcessedTargetEntryPoint(targetPath: AbsoluteFsPath, propertiesToConsider: string[]) {
+  const packageJsonPath = AbsoluteFsPath.from(resolve(targetPath, 'package.json'));
+  const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
+  return propertiesToConsider.some(
```
Should this take into consider `compileAllFormats` when checking `propertiesToConsider`?

Also, the logic might not be always correct. If propertiesToConsider is `[main, module]` and module has been already processed. It will return `true`, even though.

petebacondarwin replied:
Yes you are right. It needs to account for `compileAllFormats`.

But if that is set to false then if any of the `propertiesToConsider` has been processed I would argue that we are done. Even if it is not the first one?

alan-agius4 replied:
Not quite, in propertiesToConsider we are passing the fields to consider sorted by priority and fall backing ordering.

If we pass [module, main], it should use main only in case when a library doesn’t have a module. Hence this means that it needs to be process by ngcc.

The main reason for this is that we don’t for example want to end up with umd code in an es5 app just because ngcc process the umd bundle first.

alan-agius4 replied:
Ps: we can put this in the list of things to discuss on Monday 😜

petebacondarwin replied:
I think this will cover us:

```ts
function hasProcessedTargetEntryPoint(
    targetPath: AbsoluteFsPath, propertiesToConsider: string[], compileAllFormats: boolean) {
  const packageJsonPath = AbsoluteFsPath.from(resolve(targetPath, 'package.json'));
  const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
  for (const property of propertiesToConsider) {
    if (packageJson[property]) {
      // There is a property that should be processed
      if (hasBeenProcessed(packageJson, property as EntryPointJsonProperty)) {
        if (!compileAllFormats) {
          // It has been processed and we only need one, so we are done.
          return true;
        }
      } else {
        if (compileAllFormats) {
          // It has not been processed and we need all, so we are done.
          return false;
        }
      }
    }
    // Either all formats should be compiled and there was at least one that was unprocessed.
    // Or only the first format should be compiled but they were all unprocessed.
    return false;
  }
}
```

petebacondarwin replied:
Actually I don't think it does :-(

alan-agius4 replied:
Yeah, LGTM.

alan-agius4 replied:
Why not?

alan-agius4 replied:
Why not?

alan-agius4 replied:
Ah it won’t it needs a small change. It need to stop the iteration when a property exists and not processed

alan-agius4 replied:
@petebacondarwin the below should do
```ts
function hasProcessedTargetEntryPoint(targetPath: AbsoluteFsPath, propertiesToConsider: string[], compileAllFormats: boolean) {
  const packageJsonPath = AbsoluteFsPath.from(resolve(targetPath, 'package.json'));
  const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
  const properties = Object.keys(packageJson);
  // filter out propertiesToConsider that don't exists in package.json
  const availablePropertiesToConsider = propertiesToConsider.filter(x => properties.includes(x));

  if (availablePropertiesToConsider.length === 0) {
    // no properties available
    return true;
  }

  if (compileAllFormats) {
    // if any one of the properties is not processed return false
    return !availablePropertiesToConsider.some(x => !hasBeenProcessed(packageJson, x as EntryPointJsonProperty));
  } else {
    // otherwise we should consider the first property to consider. 
    return hasBeenProcessed(packageJson, availablePropertiesToConsider[0] as EntryPointJsonProperty);
  }
}
```

petebacondarwin replied:
Unfortunately that doesn't quite do it either. If `compileAllFormats === false` then we can't just check the first property. This is because that target might not contain that property, in which case we would fall back to a later property.

petebacondarwin replied:
This is where I am now:

```ts
function hasProcessedTargetEntryPoint(
    targetPath: AbsoluteFsPath, propertiesToConsider: string[], compileAllFormats: boolean) {
  const packageJsonPath = AbsoluteFsPath.from(resolve(targetPath, 'package.json'));
  const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
  for (const property of propertiesToConsider) {
    if (packageJson[property]) {
      // Here is a property that should be processed
      if (hasBeenProcessed(packageJson, property as EntryPointJsonProperty)) {
        if (!compileAllFormats) {
          // It has been processed and we only need one, so we are done.
          return true;
        }
      } else {
        // It has not been processed, so we are done.
        return false;
      }
    }
    // Either all formats need to be compiled and there were none that were unprocessed,
    // Or only the one matching format needs to be compiled but there was at least one matching
    // property before the first processed format that was unprocessed.
    return true;
  }
}
```

(Also there should be no need to check the length of `propertiesToConsider`. Why would anyone ever pass an empty array anyway?)

alan-agius4 replied:
I am already handling that scenario

```
const availablePropertiesToConsider = propertiesToConsider.filter(x => properties.includes(x));
```

alan-agius4 replied:
😂😂😂

petebacondarwin replied:
This is what I came up with, which is slightly faster - not that there would be any appreciable difference:

```ts
function hasProcessedTargetEntryPoint(
    targetPath: AbsoluteFsPath, propertiesToConsider: string[], compileAllFormats: boolean) {
  const packageJsonPath = AbsoluteFsPath.from(resolve(targetPath, 'package.json'));
  const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
  
  for (const property of propertiesToConsider) {
    if (packageJson[property]) {
      // Here is a property that should be processed
      if (hasBeenProcessed(packageJson, property as EntryPointJsonProperty)) {
        if (!compileAllFormats) {
          // It has been processed and we only need one, so we are done.
          return true;
        }
      } else {
        // It has not been processed but we need all of them, so we are done.
        return false;
      }
    }
    // Either all formats need to be compiled and there were none that were unprocessed,
    // Or only the one matching format needs to be compiled but there was at least one matching
    // property before the first processed format that was unprocessed.
    return true;
  }
}
```

alan-agius4 replied:
That should do it as well.

alan-agius4 replied:
Shouldn’t last `return true` Be outside of the loop?

petebacondarwin replied:
Oops yes! Github-driven-development :-)
I am writing tests; to cover all the cases, I hope.

petebacondarwin replied:
Updated.

petebacondarwin committed:
test(ivy): ngcc - improve and use the `MockLogger`

Previously the console logger was being used in integration tests
leading to lots of output during test runs.

petebacondarwin committed:
perf(ivy): ngcc - exit early if the targeted package has been compiled

Previously we always walked the whole folder tree looking for
entry-points before we tested whether a target package had been
processed already. This could take >10secs!

This commit does a quick check of the target package before doing
the full walk which brings down the execution time for ngcc in this
case dramatically.

```
$ time ./node_modules/.bin/ivy-ngcc -t @angular/common/http/testing
Compiling @angular/core : fesm2015 as esm2015
Compiling @angular/core : fesm5 as esm5
Compiling @angular/core : esm2015 as esm2015
Compiling @angular/core : esm5 as esm5
Compiling @angular/common/http : fesm2015 as esm2015
Compiling @angular/common/http : fesm5 as esm5
Compiling @angular/common/http : esm2015 as esm2015
Compiling @angular/common/http : esm5 as esm5
Compiling @angular/common/http/testing : fesm2015 as esm2015
Compiling @angular/common/http/testing : fesm5 as esm5
Compiling @angular/common/http/testing : esm2015 as esm2015
Compiling @angular/common/http/testing : esm5 as esm5

real	0m19.766s
user	0m28.533s
sys	0m2.262s
```

```
$ time ./node_modules/.bin/ivy-ngcc -t @angular/common/http/testing
The target entry-point has already been processed

real	0m0.666s
user	0m0.605s
sys	0m0.113s
```

alan-agius4 commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -6,15 +6,19 @@
+import {EntryPointJsonProperty, EntryPointPackageJson, SUPPORTED_FORMAT_PROPERTIES} from '../../src/packages/entry_point';
+import {MockLogger} from '../helpers/mock_logger';
+
+const _ = AbsoluteFsPath.from;
```
Why naming this `_`?

petebacondarwin replied:
In unit tests I have to create `AbsoluteFsPath`s from strings all over the place, so I got into a habit of aliasing `AbsoluteFsPath.from()` to `_()` to make the tests less verbose.

I realise  that in this file I only use it once so perhaps that isn't necessary. :-)

alan-agius4 replied:
I typically associate  _ with lodash or understand 😜.

petebacondarwin replied:
That's so 1990s :-P

Check out https://github.com/angular/angular/blob/master/packages/compiler-cli/ngcc/test/packages/dependency_host_spec.ts for an example of this being actually useful.

alan-agius4 commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -167,3 +175,28 @@ export function mainNgcc({basePath, targetEntryPointPath,
   return createNewEntryPointFormats ? new NewEntryPointFileWriter() : new InPlaceFileWriter();
 }
+
+function hasProcessedTargetEntryPoint(
```
This looks good

alan-agius4 reviewed (approved):
LGTM

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -167,3 +175,28 @@ export function mainNgcc({basePath, targetEntryPointPath,
+
+function hasProcessedTargetEntryPoint(
+    targetPath: AbsoluteFsPath, propertiesToConsider: string[], compileAllFormats: boolean) {
+  const packageJsonPath = AbsoluteFsPath.from(resolve(targetPath, 'package.json'));
```
Since you are concerned with speed. `fromUnchecked()` should be (negligibly) faster (and sufficient) :grin:

petebacondarwin replied:
I'll fix that up next time I am touching this code.

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -167,3 +175,28 @@ export function mainNgcc({basePath, targetEntryPointPath,
+    }
+  }
+  // Either all formats need to be compiled and there were none that were unprocessed,
+  // Or only the one matching format needs to be compiled but there was at least one matching
```
_Or_ --> _or_

petebacondarwin replied:
😱

gkalpak reviewed (commented):
Post-merge review, because...why not 😇
